docs/design/coreclr/botr/data_contracts_for_coreclr.md

```diff
@@ -0,0 +1,430 @@
+
+## Contract data model
+Logically a contract may refer to another contract. If it does so, it will typically refer to other contracts by names which do not include the contract version. This is to allow for version flexibility. Logically once the contract data stream is fully processed, there is a single list of contracts that represents the set of contracts useable with whatever runtime instance is being processed.
+
+## Types of contracts
+
+There are 3 different types of contracts each representing a different phase of execution of the data contract system.
+
+### Composition contracts
+These contracts indicate the version numbers of other contracts. This is done to reduce the size of contract list needed in the runtime data stream. In general it is intended that as a runtime nears shipping, the product team can gather up all of the current versions of the contracts into a single magic value, which can be used to initialize most of the contract versions of the data contract system. A specific version number in the contract data stream for a given contract will override any composition contracts specified in the contract data stream. If there are multiple composition contracts in a stream which specify the same contract to have a different version, the first composition contract linearly in the runtime data stream wins. This is intended to allow for a composite contract for the architecture/os indepedent work, and a separate composite contract for the non independent work. If a contract is specified explicitly in the Data Contract Stream and a different version is specified via the composition contract mechanism, the explicitly specified contract takes precedence.
+
+### Fixed value contracts
+These contracts represent data which is entirely determined by the contract version + contract name. There are 2 subtypes of this form of contract.
+
+#### Global Value Contract
+A global value contract specifies numbers which can be referred to by other contracts. If a global value is specified directly in the Contract Data Stream, then the global value defintion in the Contract Data Stream takes precedence. The intention is that these global variable contracts represent magic numbers and values which are useful for the operation of algorithmic contracts. For instance, we will likely have a `TargetPointerSize` global value represented via a contract, and things like `FEATURE_SUPPORTS_COM` can also be a global value contract, with a value of 1.
+
+#### Data Structure Definition Contract
+A data structure definition contract defines a single type's physical layout. It MUST be named "MyDataStructureType_layout". If a data structure layout is specified directly in the Contract Data Stream, then the data structure defintion in the Contract Data Stream takes precedence. These contracts are responsible for declaring the field layout of individual fields.
```

Comment:
> If a data structure layout is specified directly in the Contract Data Stream, then the data structure defintion in the Contract Data Stream takes precedence.

Trying to make sure I understand the precedence/override mechanism described here. So if the stream/descriptor has records:
```
[compatible contracts]
Composition1: v1
    documented as [ Data1_layout : v1, Data2_layout: v1, Data3_layout: v1, Algorithm1 : v3 ]
Composition2: v1
    documented as [ Data1_layout : v3, Data3_layout: v2, Algorithm2 : v1 ]
Data1_layout : v2
Algorithm1 : v2

[data structure layout]
Data2 : <fields>
```

The result should be the below? (And the same idea for global values)
```
Data1_layout : v2        // Data1_layout compatible contract record (precedence over Composition1 and Composition2)
Data2_layout : <fields>  // Data2 data structure layout record (precedence over any compatible contract record)
Data3_layout : v1        // Composition1 compatible contract record (precedence over Composition2)
Algorithm1 : v2          // Algorithm1 compatible contract record (precedence over Composition1)
Algorithm2 : v1          // Composition2 compatible contract record
```

Reply:
As currently framed I am also unclear what precedence rules are intended.

I'm hoping we avoid needing composition contracts at all, but if we do have them my next hope is that we can avoid having any override/precedence mechanics. We'd require that any given contract only has its version implied at most once, either by listing it explicitly or by including it in at most one composition.

Reply:
@elinor-fung interprets my inscrutable prose correctly. However, we can discuss whether or not precedence rules make any sense. @noahfalk, I'd like your opinion on my rationales here.

 So there are really 2 different purposes for the precedence rules.

1. To allow us to depend on a composition contracts + a delta that is what the runtime actually implements. (For instance, we have a composition contract which is the behavior of .NET 9 + v2 of GCWalking, because we started to change that early in .NET 10. Based on the feedback from folks in this PR, I expect to remove composition contracts from this version of the spec, and we can add it back later if it turns out to be desirable.

2. To allow us to define the data structure layouts into the Data Contract Reader explicitly only for release formats, but for DEBUG or other one-off builds, that we would have a means to allow for data structures to have significantly different sets of offsets, and yet still be conveniently debuggable. For instance, one of the things that I sometimes do is add a field to a common data structure with a string that describes what it is, what is going on, etc. The idea is to allow that sort of ad hoc hackery to not effect the debuggability of the runtime generated.

Reply:
Thanks David! My preference remains to avoid composition contracts if possible, but hypothetically if we were to have them then I think the next best choices would be:
scenario 1 - At the time we update the GCWalking contract, also update the composition which contains it. Hopefully we've defined the compositions such that the debugger functionality lost when breaking a member isn't that different than what is lost breaking the entire composition. GCWalking sounds like a relatively large contract already. I am assuming if we have composition then we have some finer-grained contracts composing it like GCInfoCompositionContract = Segment + Generation + GCHeap + AllocContext + GCGlobals. If you made a breaking change to Segment then I assume you've already broken most of the GC debugging functionality regardless.
Scenario 2 - I think it might be a good practice for contracts to specify default values of field offsets in documentation, but also include some place where optional explicit offsets can be included with the encoded contract data. If we had composition contracts then I'd propose they should support the same parameterization. If we decided that the best way to encode optional parameters of composition contracts was to explicitly encode some member contracts I think that would be OK as long as both the composition and the explicit member contract agree on the same version. This is a little more relaxed than my initial thought above.